File: backend/src/Operum.API/Controllers/CorrelationInsightsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Operum.API.Controllers.Base;
using Operum.Service.Interfaces;

namespace Operum.API.Controllers
{
    [Authorize]
    [ApiController]
    [Route("api/[controller]")]
    public class CorrelationInsightsController(ICorrelationInsightService correlationInsightService) : BaseController
    {
        [HttpGet]
        public async Task<IActionResult> GetInsights()
        {
            return GetApiResponse(await correlationInsightService.GetInsights());
        }

        // Recomputes synchronously rather than queuing a job: a user's own tracker data is
        // small enough that this finishes well within a normal request.
        [HttpPost("run")]
        public async Task<IActionResult> RunNow(CancellationToken ct)
        {
            return GetApiResponse(await correlationInsightService.RunNow(ct));
        }

        [HttpPost("{insightId}/dismiss")]
        public async Task<IActionResult> Dismiss([FromRoute] string insightId)
        {
            return GetApiResponse(await correlationInsightService.Dismiss(insightId));
        }
    }
}

File: frontend/src/features/widgets/api/correlationInsightsController.ts
import api from "../../../shared/api/api";
import { ApiResponse } from "../../../shared/types/ApiResponse";
import { CorrelationInsightDto } from "../types/CorrelationInsightDto";

export const correlationInsightsController = {
    getInsights: async (): Promise<ApiResponse<CorrelationInsightDto[]>> => {
        return await api.get("/correlationinsights");
    },

    runNow: async (): Promise<ApiResponse<CorrelationInsightDto[]>> => {
        return await api.post("/correlationinsights/run");
    },

    dismiss: async (insightId: string): Promise<ApiResponse> => {
        return await api.post(`/correlationinsights/${insightId}/dismiss`);
    },
};
